packages/fuels-types/src/param_types.rs

```diff
@@ -88,110 +96,273 @@ impl ParamType {
+impl Type {
+    /// Will recursively drill down the given generic parameters until all types are
+    /// resolved.
+    ///
+    /// # Arguments
+    ///
+    /// * `type_application`: the type we wish to resolve
+    /// * `types`: all types used in the function call
+    pub fn from(
+        type_application: &TypeApplication,
+        type_lookup: &HashMap<usize, TypeDeclaration>,
+    ) -> Self {
+        Self::resolve(type_application, type_lookup, &[])
+    }
+
+    fn resolve(
+        type_application: &TypeApplication,
+        type_lookup: &HashMap<usize, TypeDeclaration>,
+        parent_generic_params: &[(usize, Type)],
+    ) -> Self {
+        let type_decl = type_lookup.get(&type_application.type_id).unwrap();
+
+        if extract_generic_name(&type_decl.type_field).is_some() {
+            let (_, generic_type) = parent_generic_params
+                .iter()
+                .find(|(id, _)| *id == type_application.type_id)
+                .unwrap();
+
+            return generic_type.clone();
+        }
+
+        // Figure out what does the current type do with the inherited generic
+        // parameters and reestablish the mapping since the current type might have
+        // renamed the inherited generic parameters.
+        let generic_params_lookup =
+            Self::determine_generics_for_type(type_application, type_lookup, parent_generic_params);
+
+        // Resolve the enclosed components (if any) with the newly resolved generic
+        // parameters.
+        let components = type_decl
+            .components
+            .iter()
+            .flatten()
+            .map(|component| Self::resolve(component, type_lookup, &generic_params_lookup))
+            .collect_vec();
+
+        Type {
+            type_field: type_decl.type_field.clone(),
+            components,
+            generic_params: generic_params_lookup
+                .into_iter()
+                .map(|(_, ty)| ty)
+                .collect(),
+        }
+    }
+
+    /// For the given type generates generic_type_id -> Type mapping describing to
+    /// which types generic parameters should be resolved.
+    ///
+    /// # Arguments
+    ///
+    /// * `type_application`: The type on which the generic parameters are defined.
+    /// * `types`: All types used.
+    /// * `parent_generic_params`: The generic parameters as inherited from the
+    ///                            enclosing type (a struct/enum/array etc.).
+    fn determine_generics_for_type(
+        type_application: &TypeApplication,
+        types: &HashMap<usize, TypeDeclaration>,
+        parent_generic_params: &[(usize, Type)],
+    ) -> Vec<(usize, Self)> {
+        let type_decl = types.get(&type_application.type_id).unwrap();
+        match &type_decl.type_parameters {
+            // The presence of type_parameters indicates that the current type
+            // (a struct or an enum) defines some generic parameters (i.e.
+            // SomeStruct<T, K>
```

Comment:
Missing matching parenthesis:

```suggestion
            // The presence of type_parameters indicates that the current type
            // (a struct or an enum) defines some generic parameters 
            // (i.e. SomeStruct<T, K>).
```